snuba/query/processors/logical/filter_in_select_optimizer.py

```diff
@@ -0,0 +1,188 @@
+            return self._get_domain_of_function_call(exp)
+        elif isinstance(exp, CurriedFunctionCall):
+            return self._get_domain_of_curried_function_call(exp)
+        elif isinstance(exp, Literal):
+            return set()
+        else:
+            # we dont know so we just say its None
+            return None
+
+    def _get_domain_of_function_call(self, f: FunctionCall) -> Domain:
+        """
+        The domain of a function call is either:
+            - case 1: its a conditional aggregate, so the domain of that
+            - case 2: any other function is the union of domains of its params
+                ex. sumIf1 + sumIf2 / sumIf3
+        """
+        if f.function_name[-2:] == "If":
+            predicate = f.parameters[1]
+            assert isinstance(predicate, FunctionCall)
+            return self._get_domain_of_predicate(predicate)
+        else:
+            if len(f.parameters) == 0:
+                return None
+            else:
+                domain = self._get_domain_of_exp(f.parameters[0])
+                for i in range(1, len(f.parameters)):
+                    domain = self._domain_union(
+                        domain, self._get_domain_of_exp(f.parameters[i])
+                    )
+                return domain
+
+    def _get_domain_of_curried_function_call(self, f: CurriedFunctionCall) -> Domain:
+        if f.internal_function.function_name[-2:] == "If":
+            assert isinstance(f.parameters[1], FunctionCall)
+            return self._get_domain_of_predicate(f.parameters[1])
+        else:
+            return None
+
+    def _get_domain_of_predicate(self, predicate: FunctionCall) -> set[int] | None:
+        if predicate.function_name == "equals":
+            return self._get_domain_of_equals(predicate)
+        elif predicate.function_name == "in":
+            return self._get_domain_of_in(predicate)
+        elif predicate.function_name == "and":
+            return self._get_domain_of_and(predicate)
+        else:
+            return None
+
+    def _get_domain_of_equals(self, e: FunctionCall) -> set[int] | None:
+        assert e.function_name == "equals"
+        lhs = e.parameters[0]
+        if isinstance(lhs, Column) and lhs.column_name == "metric_id":
+            rhs = e.parameters[1]
+            assert isinstance(rhs, Literal)
+            assert isinstance(rhs.value, int)
+            return {rhs.value}
+        else:
+            return None
+
+    def _get_domain_of_in(self, e: FunctionCall) -> set[int] | None:
+        assert e.function_name == "in"
+        lhs = e.parameters[0]
+        rhs = e.parameters[1]
+        if isinstance(lhs, Column) and lhs.column_name == "metric_id":
+            assert isinstance(rhs, FunctionCall) and rhs.function_name == "array"
+            ids = set()
+            for metric_id in rhs.parameters:
+                assert isinstance(metric_id, Literal) and isinstance(
+                    metric_id.value, int
+                )
```

Comment:
A corner case I could foresee here is something like

```
in(1,2,3,4, sum(1, 2))
```
the example may be silly (people don't add metric ids) but overall it seems like here you will fail the query entirely if there's an `in` condition with a functioncall as a parameter even though the query is syntactically valid. Better thing to do in such a case would be to not apply the optimization.

Reply:
To add to this: your functions in general have a lot of `assert` statements. Generally we only have `assert` statements in our code to satisfy typing (and in tests). There is an assumption that an assert should never be thrown.

It's hard to follow the logic of your code to see that an assert statement is guaranteed to pass. So in this example, it seems like there are cases where the assertion would fail. Instead of throwing an assertion error, that case should be handled either by throwing a more descriptive error, or bailing out of the optimization (probably the latter).